FirebaseRemoteConfig/SwiftNew/ConfigExperiment.swift

```diff
@@ -0,0 +1,192 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import FirebaseABTesting
+import Foundation
+
+// TODO(ncooke3): Once everything is ported, the `@objc` and `public` access
+// can be removed.
+
+/// Handles experiment information update and persistence.
+@objc(RCNConfigExperiment) public final class ConfigExperiment: NSObject {
+  private static let experimentMetadataKeyLastStartTime = "last_experiment_start_time"
+  private static let serviceOrigin = "frc"
+
+  @objc private var experimentPayloads: [Data]
+  @objc private var experimentMetadata: [String: Any]
+  @objc private var activeExperimentPayloads: [Data]
+  private let dbManager: ConfigDBManager?
+  private let experimentController: ExperimentController?
+  private let experimentStartTimeDateFormatter: DateFormatter
+
+  /// Designated initializer;
+  @objc public init(DBManager: ConfigDBManager?,
+                    experimentController controller: ExperimentController?) {
+    experimentPayloads = []
+    experimentMetadata = [:]
+    activeExperimentPayloads = []
+    experimentStartTimeDateFormatter = {
+      let dateFormatter = DateFormatter()
+      dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"
+      // Locale needs to be hardcoded. See
+      // https://developer.apple.com/library/ios/#qa/qa1480/_index.html for more details.
+      dateFormatter.locale = Locale(identifier: "en_US_POSIX")
+      dateFormatter.timeZone = TimeZone(abbreviation: "UTC")
+      return dateFormatter
+    }()
+    dbManager = DBManager
+    experimentController = controller
+    super.init()
+    loadExperimentFromTable()
+  }
+
+  @objc private func loadExperimentFromTable() {
+    guard let dbManager else { return }
+
+    let completionHandler: (Bool, [String: Any]?) -> Void = { [weak self] _, result in
+      guard let self else { return }
+
+      if result?[ConfigConstants.experimentTableKeyPayload] != nil {
+        self.experimentPayloads.removeAll()
+        if let experiments = result?[ConfigConstants.experimentTableKeyPayload] as? [Data] {
+          for experiment in experiments {
+            do {
+              try JSONSerialization.jsonObject(with: experiment)
+              self.experimentPayloads.append(experiment)
+            } catch {
+              RCLog.warning("I-RCN000031", "Experiment payload could not be parsed as JSON.")
+            }
+          }
+        }
+      }
+
+      if let experimentTable =
+        result?[ConfigConstants.experimentTableKeyMetadata] as? [String: Any] {
+        self.experimentMetadata = experimentTable
+      }
```

Comment:
@paulb777, just calling this out following 475a6799533e6445d66127343cdf8c56290eade2, making the property non-optional means we need to unwrap and cast the new value. The ObjC implementation would do the nil check but it didn't do the casting. In Swift, we have to cast it the dictionary value to the expected type, `[String: Any]`, and that means we have an extra nil case to handle. In the event the cast fails, we don't update the property. This is a slight deviation from the ObjC impl. as the an value of an unexpected type would be assigned to the property anyway. I think this is a reasonable deviation since I'd imagine this situation in the ObjC implementation to be very unlikely and it would likely cause unexpected behavior at some point afterwards.